autofit/message_passing/messages/abstract.py

```diff
@@ -0,0 +1,256 @@
+        pass
+
+    @property
+    def scale(self):
+        return self.variance ** 0.5
+
+    def __init__(self, parameters, log_norm=0.):
+        self.log_norm = log_norm
+        self._broadcast = np.broadcast(*parameters)
+        self.parameters = parameters
+
+    def __iter__(self):
+        return iter(self.parameters)
+
+    @property
+    def shape(self):
+        return self._broadcast.shape
+
+    @property
+    def ndim(self):
+        return self._broadcast.ndim
+
+    @classmethod
+    def from_natural_parameters(cls, parameters, **kwargs):
+        args = cls.invert_natural_parameters(parameters)
+        return cls(*args, **kwargs)
+
+    @classmethod
+    @abstractmethod
+    def invert_sufficient_statistics(cls, sufficient_statistics):
+        pass
+
+    @classmethod
+    def from_sufficient_statistics(cls, suff_stats, **kwargs):
+        natural_params = cls.invert_sufficient_statistics(suff_stats)
+        return cls.from_natural_parameters(natural_params, **kwargs)
+
+    def sum_natural_parameters(self, *dists):
+        """return the unnormalised result of multiplying the pdf
+        of this distribution with another distribution of the same
+        type
+        """
+        log_norm = self.log_norm + sum(
+            dist.log_norm for dist in self._iter_dists(dists))
+        new_params = sum(
+            (dist.natural_parameters for dist in self._iter_dists(dists)),
+            self.natural_parameters)
+        mul_dist = self.from_natural_parameters(new_params, log_norm=log_norm)
+        return mul_dist
+
+    def sub_natural_parameters(self, other):
+        """return the unnormalised result of dividing the pdf
+        of this distribution with another distribution of the same
+        type"""
+        log_norm = self.log_norm - other.log_norm
+        new_params = (self.natural_parameters - other.natural_parameters)
+        div_dist = self.from_natural_parameters(new_params, log_norm=log_norm)
+        return div_dist
+
+    _multiply = sum_natural_parameters
+    _divide = sub_natural_parameters
+
+    def __mul__(self, other):
+        if np.isscalar(other):
+            log_norm = self.log_norm + np.log(other)
+            return type(self)(*self.parameters, log_norm=log_norm)
+        else:
+            return self._multiply(other)
+
+    def __rmul__(self, other):
+        return self * other
+
+    def __truediv__(self, other):
+        if np.isscalar(other):
+            log_norm = self.log_norm - np.log(other)
+            return type(self)(*self.parameters, log_norm=log_norm)
+        else:
+            return self._divide(other)
+
+    def __pow__(self, other):
+        natural = self.natural_parameters
+        new_params = other * natural
+        log_norm = other * self.log_norm
+        return self.from_natural_parameters(new_params, log_norm=log_norm)
+
+    def __repr__(self):
+        return f"{type(self).__name__}({self}"
+
+    def __str__(self):
+        return ""
```

Comment:
Why did you remove the formatting of the repr? It can be quite useful when debugging in interactive mode? I'm guessing the linter didn't like it?

Reply:
Good question. I think I actually meant to fix it afterwards but it got in the way of refactoring something else. Will see if I can put it back